src/main/kotlin/org/rust/ide/inspections/RsSelfConventionInspection.kt

```diff
@@ -51,27 +62,55 @@ class RsSelfConventionInspection : RsLocalInspectionTool() {
     }
 }
 
-enum class SelfSignature(val description: String) {
-    NO_SELF("no self"),
-    BY_VAL("self by value"),
-    BY_REF("self by reference"),
-    BY_MUT_REF("self by mutable reference");
+sealed class SelfSignature {
+    object ArbitrarySelfSignature: SelfSignature()
+    class BasicSelfSignature(val description: String): SelfSignature()
+
+    companion object {
+        val NO_SELF: BasicSelfSignature = BasicSelfSignature("no self")
+        val BY_VAL: BasicSelfSignature = BasicSelfSignature("self by value")
+        val BY_REF: BasicSelfSignature = BasicSelfSignature("self by reference")
+        val BY_MUT_REF: BasicSelfSignature = BasicSelfSignature("self by mutable reference")
+    }
 }
 
 private val RsFunction.selfSignature: SelfSignature
     get() {
-        val self = selfParameter
-        return when {
-            self == null -> SelfSignature.NO_SELF
-            self.isRef && self.mutability.isMut -> SelfSignature.BY_MUT_REF
-            self.isRef -> SelfSignature.BY_REF
-            else -> SelfSignature.BY_VAL
+        val self = selfParameter ?: return SelfSignature.NO_SELF
+
+        // self: Self, self: &Self, self: Box<Self>, ...
+        return if (self.isExplicitType) {
+            val expectedSelfType = selfType ?: TyUnknown
+            val actualSelfType = self.typeOfValue
```

Comment:
I think this can be done a bit simple by using `self.typeOfValue(TyTypeParameter.self())` and then using simple comparison with `TyTypeParameter.self()` instead of `expectedSelfType`.

Reply:
Did you mean it like this?
```kotlin
return if (self.isExplicitType) {
            when (val actualSelfType = self.typeOfValue(TyTypeParameter.self())) {
                TyTypeParameter.self() -> SelfSignature.BY_VAL
                is TyReference -> {
                    val mutable = actualSelfType.mutability.isMut
                    if (actualSelfType.referenced == TyTypeParameter.self()) {
                        if (mutable) {
                            SelfSignature.BY_MUT_REF
                        } else {
                            SelfSignature.BY_REF
                        }
                    } else {
                        null
                    }
                }
                else -> null
```
the tests fail with this, I didn't investigate further yet.

Reply:
Yes, I thought it would work.

Reply:
(can I still merge this or do you want me to investigate? :D )